utils/weather_tools.py

```diff
@@ -37,114 +23,230 @@ def generate_stationID_dict():
-    try:
-        df = pd.read_csv(url_ftp, compression='gzip', header=None)
-        df['TEMPERATURE' + stationID] = df[0].apply(extract_Temp)
-        df['Date'] = df[0].apply(extract_date)
-        df.dropna(inplace=True)
-        df.drop(columns=[0], inplace=True)
-        df['Date'] = pd.to_datetime(df['Date'])
-        df['TEMPERATURE' + stationID] = df['TEMPERATURE' +
-                                           stationID].astype('int64')
-        df.reset_index(inplace=True)
-        df.drop(columns=['index'], inplace=True)
-        return df
-    except (pd.errors.ParserError, OSError, ValueError) as e:  # URLError:
-        #print('No good')
-        print(f'{e} with file: ' + url_ftp)
-        Date = pd.date_range(str(year) + '-01-01', str(year) + '-12-31')
-        df = pd.DataFrame(data=Date, columns=['Date'])
-        return df
-
+key_list = df_stations_US.StID.values
+valu_list = df_stations_US.TZone.values 
+TZ_dict = {key_list[i] : valu_list[i] for i in range(len(key_list))} # dictionary containing {Station_ID: Time-Zone}
+
+def process_file(f):
+    
+    StID = f[-20:-8]
+    good = ['1','5', 'A', 'C', 'U'] #temperature quality codes which indicate good data
+    df = pd.read_csv(f, names = ['data'], compression='gzip', header=None)
+    df_no_summary = df[df.data.str.contains('FM-12')].copy()
+    if len(df_no_summary) != 0: #check for files not containing FM-12
+        df_no_summary.reset_index(drop = True, inplace = True)
+        df_no_summary['Date'] = pd.to_datetime(df_no_summary['data'].str[15:27], utc = True)
+        df_no_summary['Date'] = df_no_summary.apply(lambda x: x.Date.astimezone(tz=TZ_dict[StID]), axis = 1) #converting to local timezone
+        df_no_summary['T_C'] = df_no_summary.apply(lambda row : int(row.data[87:92])/10 if (row.data[92] in good) else np.nan, axis = 1)    
+        df_no_summary.drop(columns = ['data'], inplace = True) 
+        return df_no_summary.resample('D', on = 'Date').mean()
+
+    else: #skip file. Does not contain FM-12
+        return 0
 
 def all_stations_for_state_year(state, year):
     """Reads data from all stations in a given state and year
 
-    Returns a dataframe with all the dates in that year temperatures. 
-    The last column contains mean temperatures.
+    Returns: DataFrame with Date index and separate columns for each station-ID
+    
     """
 
     df_list = []
-    Date = pd.date_range(str(year) + '-01-01', str(year) + '-12-31')
-    df = pd.DataFrame(data=Date, columns=['Date'])
-    for stID in stations_in_state[state]:
-        # print(stID)
-        xx = df_from_file(stID, year)
-        # put code here which gets rid of leap years?
-        df_list.append(xx)
+
+    files_processed = 0
+    not_found = 0
+    files_wo_FM12 = 0
+
+    tot_files = len(stations_in_state[state])
+    print(f"Files to process: {tot_files}")
+
+    for StID in stations_in_state[state]:
+        f = "/Volumes/Nikhil hdd/Weather Data/" + year + "/" + StID + "-" + year + ".gz"
```

Comment:
Best practice here might be to define a more generic weather data directory, like (just as an example)
```
weather_dir = '/Volumes/Nikhil hdd/Weather Data/'
f = weather_dir + year  # etc
```.

However, I understand this was a huge amount of data that we are  _hoping_ to not have to download again.